ueberzug/X/display.c

```diff
@@ -92,25 +93,50 @@ has_property(DisplayObject *self, Window window, Atom property) {
     return status == Success && !(actual_type_return == None && actual_format_return == 0);
 }
 
+static void
+get_child_window_ids_helper(Display *display, Window window,
+        Window **total_children, unsigned int *total_num_children) {
+    Window _, *children;
+    unsigned int num_children;
+
+    XQueryTree(display, window, &_, &_, &children, &num_children);
+    if (!children) return;
+
+    if (!*total_children) {
+        *total_children = calloc(num_children, sizeof(Window));
+    } else {
+        Window *tmp = reallocarray(*total_children, num_children + *total_num_children, sizeof(Window));
+        if (tmp) {
+            *total_children = tmp;
+        }
+    }
+
+    for (int i = 0; i < num_children; ++i) {
+        (*total_children)[i + *total_num_children] = children[i];
+    }
+    *total_num_children += num_children;
+
+    for (int i = 0; i < num_children; ++i) {
+        get_child_window_ids_helper(display, children[i], total_children, total_num_children);
+    }
+
+    XFree(children);
+}
 
 static PyObject *
 Display_get_child_window_ids(DisplayObject *self, PyObject *args, PyObject *kwds) {
     static char *kwlist[] = {"parent_id", NULL};
     Window parent = XDefaultRootWindow(self->info_display);
-    Window _, *children;
-    unsigned int children_count;
+    Window *children = NULL;
+    unsigned int children_count = 0;
 
     if (!PyArg_ParseTupleAndKeywords(
             args, kwds, "|k", kwlist,
             &parent)) {
         Py_RETURN_ERROR;
     }
 
-    if (!XQueryTree(
-            self->info_display, parent,
-            &_, &_, &children, &children_count)) {
-        raise(OSError, "failed to query child windows of %lu", parent);
-    }
+    get_child_window_ids_helper(self->info_display, parent, &children, &children_count);
```

Comment:
The `children` array is being passed to XFree later down, which seems like a wrong thing to do because it wasn't allocated via XLib. (It will *happen* to work because XFree calls `free` internally, but I don't think we should depend on XLib's internal behaviors).

Reply:
fair enough